internal/environments/models.go

```diff
@@ -0,0 +1,110 @@
+	"gorm.io/gorm"
+	// "encoding/json" // TODO: deal with the jsonb
+)
+
+// ErrEnvironmentNotFound is the error to represent a failed lookup of a environment db record
+var ErrEnvironmentNotFound = gorm.ErrRecordNotFound
+
+// dataStore is a wrapper around a gorm postgres client
+// which can be used to implement the environmentRepository interface
+type dataStore struct {
+	*gorm.DB
+}
+
+// Environment is the data structure that models a persisted to a database via gorm
+type Environment struct {
+	ID          int
+	Name        string
+	IsPermanent bool
+	Requester   string
+	Namespace   string
+	//Customization    postgres.Jsonb
+	//ClusterID int
+	//Cluster          Cluster
+	//AllocationPoolID int
+	//AllocationPool   AllocationPool
+	DestroyedAt time.Time
+	CreatedAt   time.Time
+	UpdatedAt   time.Time
+}
+
+// environmentStore is the interface defining allowed db actions for Environment
+type environmentStore interface {
+	listAll() ([]Environment, error)
+	createNew(CreateEnvironmentRequest) (Environment, error)
+	getByID(int) (Environment, error)
+	getByName(string) (Environment, error)
+}
+
+// creates a db connection via gorm
+func newEnvironmentStore(dbconn *gorm.DB) dataStore {
+	return dataStore{dbconn}
+}
+
+// CreateEnvironmentRequest struct defines the data required to create a new environment in db
+type CreateEnvironmentRequest struct {
+	Name string `json:"name" binding:"required"`
+}
+
+// creates a environment entity object to be persisted with the database from a
+// request to create a environment
+func (createEnvironmentRequest CreateEnvironmentRequest) environmentReq() Environment {
+	return Environment{
+		Name: createEnvironmentRequest.Name,
+	}
+}
+
+// db methods
+
+// Returns ALL Environments in the db
+func (db dataStore) listAll() ([]Environment, error) {
+	environments := []Environment{}
+
+	err := db.Find(&environments).Error
+	if err != nil {
+		return []Environment{}, fmt.Errorf("error retreiving environments: %v", err)
+	}
+
+	return environments, nil
+}
+
+// Saves an Environment object to the db, returns the object if successful, nil otherwise
+func (db dataStore) createNew(newEnvironmentReq CreateEnvironmentRequest) (Environment, error) {
+	newEnvironment := newEnvironmentReq.environmentReq()
+
+	if err := db.Create(&newEnvironment).Error; err != nil {
+		return newEnvironment, fmt.Errorf("error saving to database: %v", err)
+	}
+	return newEnvironment, nil
+}
+
+// Get is used to retrieve a specific environment entity from a postgres database using
+// id (primary key) as the lookup mechanism
+func (db dataStore) getByID(id int) (Environment, error) {
+	environment := Environment{}
+
+	if err := db.First(environment, id).Error; err != nil {
+		return environment, err
+	}
+	return environment, nil
+}
+
+// get a Environment by name or err otherwise
+func (db dataStore) getByName(name string) (Environment, error) {
+	environment := Environment{}
+
+	if err := db.First(environment, name).Error; err != nil {
```

Comment:
```suggestion
	if err := db.First(environment, name).Error; err != nil {
```
```suggestion
	if err := db.Where(&Environment{Name: name}).First(&environment).Error; err != nil {
```

Sorry random comment I just stumbled across this. I think that syntax is some syntactic sugar that only works for get by ID. I think other look ups need to be done like this.